### ISSUE TITLE
support GAN training procedure?

### ISSUE THREAD

**Issue by ericxsun** [ludwig-ai/ludwig#1013]
**Is your feature request related to a problem? Please describe.**

The process of training a GAN could be:

- step 1, the discriminator trains for one or more epochs
- step2, the generator trains for one or more epochs
- step3, repeat step 1 and step 2 to continue to train the generator and discriminator networks

It is quite different from the ECD (Encoder-Combiner-Decoder) procedure. So how could we train a GAN using ludwig? If not support now, what can I do if I really want it (any help or insight will be highly appreciated)

**Comment by tgaddair** [comment]
Hey @ericxsun, great suggestion!  I think Ludwig definitely should support more elaborate training procedures like GANs, it's just a question of what the right user experience is in these situations.

If we take the generator and discriminator individually, each of them can be represented as distinct Ludwig model definition.  The complexity then lies in the training procedure itself.  Specifically, we would need support for such things as:

- Multiple models per config
- Multiple optimizers (e.g., a map of `model_name: optimizer`)
- Policy for switching between training the different models
- Discriminator: augment training dataset with outputs of generator (label=fake)
- Generator: custom loss function that incorporates output of the Discriminator
- Generator: ability to train on synthetic (random noise) data

I think these would all be good general improvements to the flexibility of Ludwig.  Let me know if you think I'm missing anything.

**Comment by tgaddair** [comment]
One other thing worth mentioning is that something like this should be doable programmatically using the [train_online](https://github.com/uber/ludwig/blob/master/ludwig/api.py#L530) API today.  Something like:

```
generator = LudwigModel(generator_config)
discriminator = LudwigModel(discriminator_config)

real_samples = load_dataset()

for epoch in epochs:
    synthetic_dataset = generate_random_noise()
    generated_samples = generator.train_online(synthetic_dataset)

    combined_dataset = merge(generated_samples, real_samples)
    discriminator.train_online(combined_dataset)
```

Using this approach, the only missing piece at the moment is the custom loss function for the generator, but this should be solvable.

**Comment by ericxsun** [comment]
That's great, I'll take a look. Thanks a lot.

**Comment by w4nderlust** [comment]
i second @tgaddair , although there is a caveat, which is the fact tht at the moment there is no support in Ludwig for gradients flowing between two instances of LudwigModel. This means that when you train the discriminator the generator would not be updated.
We have to thing about ways to make that happen that are potentially impactful on the internals.
Ther is also another alternative we are considering, which is allowing for more customization of the actual model part of a Ludwig pipeline. We believe this change would enable GAN like training procedures, although it will come no earlier that a couple releases.
